core/google/api/core/retry.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2017 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for retrying functions with exponential back-off."""
+
+import datetime
+import logging
+import random
+import time
+
+import six
+
+from google.api.core import exceptions
+from google.api.core.helpers import datetime_helpers
+
+_LOGGER = logging.getLogger(__name__)
+_DEFAULT_JITTER_AMOUNT = 0.2
+
+
+def if_exception_type(*exception_types):
+    """Creates a predicate to check if the exception is of a given type.
+
+    Args:
+        exception_types (Sequence[type]): The exception types to check for.
+
+    Returns:
+        Callable[Exception]: A predicate that returns True if the provided
+            exception is of the given type(s).
+    """
+    def inner(exception):
+        """Bound predicate for checking an exception type."""
+        return isinstance(exception, exception_types)
+    return inner
+
+
+# pylint: disable=invalid-name
+# Pylint sees this as a constant, but it is also an alias that should be
+# considered a function.
+if_transient_error = if_exception_type((
+    exceptions.InternalServerError,
+    exceptions.TooManyRequests))
+"""A predicate that checks if an exception is a transient API error."""
```

Comment:
Can you add some more content:

```
"""A predicate that checks if an exception is a transient API error.

The following server errors are considered transient:

- ``InternalServerError`` (HTTP 500, gRPC ``INTERNAL(13)``)
- ``TooManyRequests`` (HTTP 429)
- ``ResourceExhausted`` (gRPC ``RESOURCE_EXHAUSTED(8)``)
"""
```

Not sure if you want to leave these out, but they are also ["equivalent"][1] to 500:

```
- ``Unknown`` (gRPC ``UNKNOWN(2)``)
- ``DataLoss`` (gRPC ``DATA_LOSS(15)``)
```

[1]: https://github.com/googleapis/googleapis/blob/660dd12452d36df708172ea01540a83eb2a4362d/google/rpc/code.proto